okio/src/main/java/okio/Timeout.java

```diff
@@ -149,4 +150,79 @@ public void throwIfReached() throws IOException {
       throw new InterruptedIOException("deadline reached");
     }
   }
+
+  /**
+   * Waits on {@code monitor} until it is notified. Throws {@link SocketTimeoutException} if either
+   * the thread is interrupted or if this timeout elapses before {@code monitor} is notified. The
+   * caller must be synchronized on {@code monitor}.
+   *
+   * <p>Here's a sample class that uses {@code waitUntilNotified()} to await a specific state. Note
+   * that the call is made within a loop to avoid unnecessary waiting and to mitigate spurious
+   * notifications. <pre>{@code
+   *
+   *   class Dice {
+   *     Random random = new Random();
+   *     int latestTotal;
+   *
+   *     public synchronized void roll() {
+   *       latestTotal = 2 + random.nextInt(6) + random.nextInt(6);
+   *       System.out.println("Rolled " + latestTotal);
+   *       notifyAll();
+   *     }
+   *
+   *     public void rollAtFixedRate(int period, TimeUnit timeUnit) {
+   *       Executors.newScheduledThreadPool(0).scheduleAtFixedRate(new Runnable() {
+   *         public void run() {
+   *           roll();
+   *          }
+   *       }, 0, period, timeUnit);
+   *     }
+   *
+   *     public synchronized void awaitTotal(Timeout timeout, int total)
+   *         throws SocketTimeoutException {
+   *       while (latestTotal != total) {
+   *         timeout.waitUntilNotified(this);
+   *       }
+   *     }
+   *   }
+   * }</pre>
+   */
+  public final void waitUntilNotified(Object monitor) throws SocketTimeoutException {
+    try {
+      boolean hasDeadline = hasDeadline();
+      long timeoutNanos = timeoutNanos();
+
+      if (!hasDeadline && timeoutNanos == 0L) {
+        monitor.wait(); // There is no timeout: wait forever.
+        return;
+      }
+
+      // Compute how long we'll wait.
+      long waitNanos;
+      long start = System.nanoTime();
+      if (hasDeadline && timeoutNanos != 0) {
+        long deadlineNanos = deadlineNanoTime() - start;
+        waitNanos = Math.min(timeoutNanos, deadlineNanos);
+      } else if (hasDeadline) {
+        waitNanos = deadlineNanoTime() - start;
+      } else {
+        waitNanos = timeoutNanos;
+      }
+
+      // Attempt to wait that long. This will break out early if the monitor is notified.
+      long elapsedNanos = 0L;
+      if (waitNanos > 0L) {
+        long waitMillis = waitNanos / 1000000L;
+        monitor.wait(waitMillis, (int) (waitNanos - waitMillis * 1000000L));
+        elapsedNanos = System.nanoTime() - start;
+      }
+
+      // Throw if the timeout elapsed before the monitor was notified.
+      if (elapsedNanos >= waitNanos) {
+        throw new SocketTimeoutException("timeout");
+      }
+    } catch (InterruptedException e) {
+      throw new SocketTimeoutException("interrupted");
```

Comment:
Yeah, this isn’t very good. I was cheating and treating `SocketTimeoutException ` as a `TimeoutException` which is a small lie. Still, I don’t like returning a boolean either here. This thing is trying to package up a timeout abstraction and throwing when it elapses is part of that.

What about keeping the SocketTimeoutException above, and trading this for an InterruptedIOException?